### ISSUE TITLE
Consolidating mass simultaneous requests

### ISSUE THREAD

**Issue by dmk23** [jpodwys/superagent-cache#41]
I have a case when my application generates a large number of identical HTTP GET requests from the same page (>10) - due to reliance on 3rd party components that I'm unable to fix.

When these requests are channeled through superagent / superagent cache I'm seeing that the latter is only able to serve a cached result if a request already successfully completed. Since the mass duplicate requests are launched within milliseconds from each other I'm running into a bottleneck.

Could superagent-cache be adapted to cache / consolidate in-process requests and use the first response as the result for all the following ones? Any pointers appreciated...

**Comment by marybeshaw** [comment]
Interesting problem.  I solved something similar in my app, using arrays of in-flight promises, using superagent but not superagent-cache.  I have some ideas about how it could be done, but I'm sure Joe has better ideas.

What are your code limitations?  Do you have to pass the superagent/superagent-cache object into your 3rd party component, and let it go?  Or can you add code around the actual superagent call?

**Comment by jpodwys** [comment]
Does this happen for all AJAX calls? Meaning, if you executed a delete or a put request using superagent, would these third-party libraries duplicate those too?

**Comment by jpodwys** [comment]
Check out [this branch](https://github.com/jpodwys/superagent-cache/tree/handle-duplicate-requests) to see if it's what you're wanting. I'm still not certain how I feel about this feature request so I'm not sure I'll merge it, but it'd be useful to know if this is what you're looking for.

**Comment by dmk23** [comment]
Yes this happens with AJAX GET calls. PUT / DELETE never have a reason to duplicate

I'm using a 3rd party dynamic select component which unfortunately has a bug that does not let me defer loading till user interacts with the component: https://github.com/JedWatson/react-select/issues/965 . If I have a complex edit screen with dozens of these elements every page load becomes an AJAX hailstorm.

Just checked out the new branch and this looks about right based on your last commit. I'm not sure how to test it though, my build process is based on npm/webpack and I'm not sure how to pull a specific github branch commit into that setup.

Is there a chance to get this released to npm under some alpha/beta/experimental label?

**Comment by jpodwys** [comment]
If you're pulling superagent-cache in via a package.json file then you can do this:

```
"superagent-cache": "jpodwys/superagent-cache#handle-duplicate-requests"
```

Try it out and let me know how it works.

Just to be completely transparent, I'm not a fan of releasing an update to a repo to work around a bug present in another repo. I do understand that it's unlikely this particular react-select issue will be fixed given that, at the moment, react-select has 377 open issues and 80 open pull requests.

That being said, I'm open to the idea and I do have a use case where a personal project could potentially benefit from this feature. However, in my use case duplicate requests are uncommon, generate a maximum of two requests per asset, and not having the feature doesn't break anything. It would really only save my backend a tiny number of requests and my front end an ocasional ~50ms delay. If you can think of some legitimate use cases (ie not to work around broken dependencies), please let me know.

**Comment by dmk23** [comment]
Just tested per your install instructions. Unfortunately still seeing duplicate identical requests :(

I think there could be plenty of cases where this functionality may be useful. Any application with great degree of parallelism risks running into identical tasks showing up all at the same time

Greatly appreciate all your help in any case!

**Comment by jpodwys** [comment]
What version of NPM are you on? Is this a public project so I can see what the issue is? If not, are you able to make a small open source repo I can pull down and run it in?

**Comment by dmk23** [comment]
```
[root@localhost]# npm -v
3.7.3
```
I did verify that  your code downloaded / updated and wiped the build dir to ensure full rebuild.

The project is private, but it was based on the public boilerplate with all the low level HTTP interaction layer pretty much unchanged: https://github.com/scottwoodall/django-react-template

You could just grab it "as is" and simulate the use case by something like this:
```
import http from "app/utils/http";
for (var i = 0; i < 20; i++) {
    http.get(endpoint, query, [successCb], [errorCb] );
}
```

The `http.js` file has been modified by adding the following code at the top:

```
import superagentCache from 'superagent-cache';
const superagent = superagentCache();
```

**Comment by jpodwys** [comment]
Ok I'm not sure how soon I'll have time to try that but feel free to fork my branch and tinker in the mean time.

**Comment by dmk23** [comment]
I'm feeling pretty stupid after realizing I forgot to actually enable the new duplicate handling feature before testing it. Once I've added `request.handleDuplicates(true);` to the `get` function the mass requests consolidated just as expected. Sorry for the confusion.

Thanks again for this excellent feature. While I don't see any reason why it should not be merged into the main branch (since it has to be turned on explicitly), I'm just as happy to use it "as is"

**Comment by jpodwys** [comment]
Haha glad you got it turned on! It needs some unit tests if it's going to be merged. While my code for superagent-cache itself it mostly clean and easy to work with, my unit test setup isn't very nice. I split my tests into different files several releases ago to better organize them but I think I did something wrong because they don't always run as expected since then. 

If you'd like to take a look to see how they can be improved and/or write some unit tests for this new feature it will help with getting it merged. 

Also, I'll probably rename this feature to something that makes more sense for what it does. handleDuplicates isn't obvious enough to me, but perhaps it's just one of those things you have to read the docs to understand? Any thoughts here would be great. Thanks!

**Comment by jpodwys** [comment]
Also, if desired, you can simply modify your http.js file by passing the defaults object to superagent-cache. This way you can tell it to always handle duplicates rather than having to do it for each query. Whatever works best for you.

**Comment by jpodwys** [comment]
I'm curious how adding it to your defaults object (so it's active for all requests) would impact your application. If you're able, please try it and let me know if all is still well.

**Comment by dmk23** [comment]
If I were to rename the feature I'd call it `cachePending`. This reflects the fact that this is just an extension of caching from the completed request to the pending / in-progress

My unit test setup is not very nice either - so probably not something I could help with at the moment. Too much work just on the basic feature development for my current projects :) 

This feature is not something I'd add to default objects though since it is only applicable to GETs. I just added an extra line to `get` method here https://github.com/scottwoodall/django-react-template/blob/master/frontend/src/app/utils/http.js#L22 and that addressed all my needs . I might play with other superagent-cache settings once my app runs into more challenging use cases...

**Comment by jpodwys** [comment]
I renamed the function in my latest commit to the branch and added a unit test but it's not passing 100% of the time. I think it's just that my test suite is poorly organized so there's some kind of timing issue or something.

**Comment by dmk23** [comment]
Tested your latest and it works and combines requests just fine.

As far as the timing issue, I think if your test suite fires parallel requests within microseconds there might be easily some concurrency / memory management issues that are hard to understand and debug. In a browser the delay between requests is on the millisecond scale and that gives enough time for many of such issues to settle. I don't think for this feature it has to be perfect to be useful... Cannot expect it to address all the shortcomings of the any given JavaScript runtime...

**Comment by jpodwys** [comment]
I apologize that this is taking me a while to decide on. Currently, I'm wondering whether this would function best as a separate superagent plugin. superagent-cache is concerned with caching requests not deduping them. Seems like it would be an interesting module.

**Comment by dmk23** [comment]
Why create a separate module and unnecessarily increase config management complexity when these are just a few optional lines of code?

In any case, this functionality probably be part of the core anyways - caching pending requests could be just as essential as caching completed requests and it isn't that different...

**Comment by jpodwys** [comment]
OK, I'm sorry this took so long. I'm ready to merge this PR. The reason I hesitated for so long is because I wasn't convinced legitimate use cases would be common and I thought that, by providing a way for people to band-aid bad architecture, I might be fostering poor application design.

However, I've since addressed both concerns. I believe a use case for this feature can occur often in various forms and I added a clear warning to the documentation for this feature.

Can you please check out the branch and let me know if it still fits your needs? If so, I'll go ahead and merge my PR.

**Comment by dmk23** [comment]
Yes, definitely would like to see it merged. Unless you've made any significant changes vs. the last time I took a look this should be good to go ...

**Comment by jpodwys** [comment]
Ok I'll merge it and cut a new release on Monday.

**Comment by jpodwys** [comment]
@dmk23 I just made another change and am now ready to merge the branch into master. Would you please first confirm that the code still works in your application before I do so?

**Comment by jpodwys** [comment]
@dmk23 Are you able to confirm that the changes work with your project?

**Comment by jpodwys** [comment]
Released in `1.7.0`.
